File: Reto # - 7/Python/PilasyColas.py
class Stack:
    def __init__(self):
        self.items = []

    def is_empty(self):
        return len(self.items) == 0

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if not self.is_empty():
            return self.items.pop()
        else:
            return None

    def peek(self):
        if not self.is_empty():
            return self.items[-1]
        else:
            return None

    def size(self):
        return len(self.items)


class Queue:
    def __init__(self):
        self.items = []

    def is_empty(self):
        return len(self.items) == 0

    def enqueue(self, item):
        self.items.append(item)

    def dequeue(self):
        if not self.is_empty():
            return self.items.pop(0)
        else:
            return None

    def peek(self):
        if not self.is_empty():
            return self.items[0]
        else:
            return None

    def size(self):
        return len(self.items)


# Ejemplo de uso
stack = Stack()
stack.push(1)
stack.push(2)
stack.push(3)
print("Pila:")
while not stack.is_empty():
    print(stack.pop())

queue = Queue()
queue.enqueue(1)
queue.enqueue(2)
queue.enqueue(3)
print("\nCola:")
while not queue.is_empty():
    print(queue.dequeue())

File: Reto # - 7/C++/PilasyColas.cpp
#include <iostream>
#include <vector>

using namespace std;

class Stack {
private:
    vector<int> elements;

public:
    void push(int item) {
        elements.push_back(item);
    }

    int pop() {
        if (isEmpty()) {
            cout << "La pila está vacía" << endl;
            return -1; // Valor de retorno por defecto en caso de pila vacía
        }
        int topElement = elements.back();
        elements.pop_back();
        return topElement;
    }

    int peek() {
        if (isEmpty()) {
            cout << "La pila está vacía" << endl;
            return -1; // Valor de retorno por defecto en caso de pila vacía
        }
        return elements.back();
    }

    bool isEmpty() {
        return elements.empty();
    }
};

class Queue {
private:
    vector<int> elements;

public:
    void enqueue(int item) {
        elements.push_back(item);
    }

    int dequeue() {
        if (isEmpty()) {
            cout << "La cola está vacía" << endl;
            return -1; // Valor de retorno por defecto en caso de cola vacía
        }
        int frontElement = elements.front();
        elements.erase(elements.begin());
        return frontElement;
    }

    int peek() {
        if (isEmpty()) {
            cout << "La cola está vacía" << endl;
            return -1; // Valor de retorno por defecto en caso de cola vacía
        }
        return elements.front();
    }

    bool isEmpty() {
        return elements.empty();
    }
};

int main() {
    // Ejemplo de uso de la pila (stack)
    Stack stack;
    stack.push(10);
    stack.push(20);
    stack.push(30);

    cout << "Elemento sacado de la pila: " << stack.pop() << endl;
    cout << "Elemento en la parte superior de la pila: " << stack.peek() << endl;

    // Ejemplo de uso de la cola (queue)
    Queue queue;
    queue.enqueue(100);
    queue.enqueue(200);
    queue.enqueue(300);

    cout << "Elemento sacado de la cola: " << queue.dequeue() << endl;
    cout << "Elemento al frente de la cola: " << queue.peek() << endl;

    return 0;
}
